Confluent kafka binding

yanmxa opened the pull request:
Signed-off-by: myan <myan@redhat.com>

Adding the Kafka binding with https://github.com/confluentinc/confluent-kafka-go. This implementation will support the following features compared to the [sarama](https://github.com/IBM/sarama) binding.

- The binding provides the ability to submit offset asynchronously. ref: https://github.com/cloudevents/sdk-go/issues/846
- Be able to subscribe to batch topics with a regular pattern. ref: https://github.com/confluentinc/confluent-kafka-go/issues/96
- Support to consume the event from a specific position(topic-partition-offset)
- Support to invoke the customized rebalance callback function for the consumer group




Note: In the confluent samples, with configuration [WithPollGoroutines(1)](https://github.com/cloudevents/sdk-go/blob/ba92a5c9a1e2508d855930049999ccd75636f0cb/v2/client/options.go#L85C6-L85C11) to ensure the order of the messages on the partition. 

   Since Kafka already natively utilizes consumer groups to enhance consumption efficiency, there is no need to use multiple goroutines to boost consumption speed. 



Resolve: https://github.com/cloudevents/sdk-go/issues/918

yanmxa commented:
/hold

embano1 commented on test/integration/kafka_confluent/kafka_test.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// To start a local environment for testing:
+// docker run --rm --net=host -e ADV_HOST=localhost -e SAMPLEDATA=0 -p 9091:9091 -p 9092:9092 lensesio/fast-data-dev
```
question: should we use the `confluent` docker images for integration testing instead? Assuming many will use this binding with `confluent` Kafka (which is Kafka, I understand, however might have some differences when it comes to defaults, connection/auth handling, etc.)

yanmxa replied:
I tested both `lensesio/fast-data-dev` and `confluentinc/confluent-local` images. I didn't find any differences in their usage during the test. 

I understand your concerns. However, running the `confluentinc/confluent-local` in the integration workflow might cause issues with the `lensesio/fast-data-dev`. I can start another PR focus on these issues. What do you think of it?

embano1 commented on v2/context/context.go:
```diff
@@ -57,6 +57,48 @@ func TopicFrom(ctx context.Context) string {
 }
 
+// Opaque key type used to store topic partition
+type topicPartitionKeyType struct{}
```
question: compared to the other types/functions in this package, this seems very Kafka specific. I'm assuming you have this in this package to influence per message handling vs in the protocol? IIRC, we have other protocol implementations allowing per-message send/receive properties in the protocol w/out changing this package. Can you please double-check?

yanmxa replied:
My initial reason for putting these properties in this package is that the sarama binding might leverage them. Like you said it's only for the kafka, to avoid unnecessary reference confusion, I have removed them from here.

yanmxa commented:
@embano1 Thanks for your review!

embano1 commented:
Thy @yanmxa and sorry for my slow responses. I'm under the water at the moment at work and can't give you a concrete date when I'll be able to review. Perhaps @duglin and @lionelvillard can take another look?

lionelvillard commented:
@yanmxa can you update the documentation: https://github.com/cloudevents/sdk-go/blob/main/docs/protocol_implementations.md?

Also I wonder if it makes sense to indicate (maybe in the documentation) this is a new implementation and it hasn't been widely tested in the wild. WDYT?

yanmxa commented:
> @yanmxa can you update the documentation: https://github.com/cloudevents/sdk-go/blob/main/docs/protocol_implementations.md?
> 
> Also I wonder if it makes sense to indicate (maybe in the documentation) this is a new implementation and it hasn't been widely tested in the wild. WDYT?

Thanks @lionelvillard! I update the protocol binding for this implementation. PTAL~

yanmxa commented:
Thanks @duglin !

yanmxa commented:
@embano1 @lionelvillard PTAL~

embano1 commented on protocol/kafka_confluent/v2/go.mod:
```diff
@@ -0,0 +1,25 @@
+replace github.com/cloudevents/sdk-go/v2 => ../../../v2
+
+require (
+	github.com/cloudevents/sdk-go/v2 v2.0.0-00010101000000-000000000000
```
bump to latest?

yanmxa replied:
Done.

embano1 commented on protocol/kafka_confluent/v2/message.go:
```diff
@@ -0,0 +1,143 @@
+
+const (
+	prefix         = "ce-"
+	contentTypeKey = "Content-Type"
```
question: why not make it lower here to directly?

yanmxa replied:
No particular reason to use the capital. I thought the `contety-type` might be a special property to decide the event format.

To align it with the sarama protocol. I changed it to lowercase.

embano1 commented on protocol/kafka_confluent/v2/message.go:
```diff
@@ -0,0 +1,143 @@
+			contentType = string(header.Value)
+		}
+		if k == specs.PrefixedSpecVersionName() {
+			contentVersion = string(header.Value)
```
question: what if `header.Value` is `1.0.2` but MUST be `1.0` according to 

> Compliant event producers MUST use a value of 1.0 when referring to this version of the specification
https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#specversion

yanmxa replied:
I‘m not sure how to handle the specversion with value 1.0.2. Maybe just validate the 'major' and 'minor' versions whether compliant with "1.0"? and if not, then print a warning message?

I checked the [other protocol](https://github.com/cloudevents/sdk-go/blob/0a3095dbbd1db5c6bbd6a039c927dcac876d4f1b/protocol/kafka_sarama/v2/message.go#L74). Then all just assert whether the specversion key exists and ignore the value. 

Do you have any other advice?

embano1 replied:
Yeah, IMHO this is a shortcoming in the other implementations. I'm fine leaving this here as is then given it's consistent with our other implementations. cc/ @duglin

duglin replied:
I like being forgiving and just ignoring the patch version number - if we want to do anything different at all. I'm ok with the current choice of only looking for 0.3 and 1.0.

embano1 commented on protocol/kafka_confluent/v2/message.go:
```diff
@@ -0,0 +1,143 @@
+)
+
+func NewMessage(msg *kafka.Message) *Message {
+	var contentType, contentVersion string
```
nit: want to add `nil` check for `msg` and `msg.TopicPartition.Topic` and `panic()` immediately if someone uses the function wrong?

yanmxa replied:
Done.

embano1 replied:
nit: just noticed code comment missing for exported function

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,158 @@
+
+var offsetKey = commitOffsetType{}
+
+// WithOffsetCtx will return the topic partitions to commit offsets for.
```
nit: this is a modify operation, so the comment seems a bit off

embano1 replied:
also, I found the name to be a bit misleading:

1) remove the `Ctx` suffix to follow option style
2) There's no offset parameter in the signature (hidden deep inside topicPartitions)

IIUC what this does it sets the offsets to start consuming from, per topic/partition. If so, rename parameter to `topicPartitionOffsets`. Also, some validations might make sense e.g., checking if all three properties are set in each slice value.

yanmxa replied:
Done

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,158 @@
+var topicPartitionKey = topicPartitionKeyType{}
+
+// WithTopicPartition returns back a new context with the given partition.
+func WithTopicPartition(ctx context.Context, partition int32) context.Context {
```
this comment is more clear than the above

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,158 @@
+// Option is the function signature required to be considered an kafka_confluent.Option.
+type Option func(*Protocol) error
+
+func WithConfigMap(config *kafka.ConfigMap) Option {
```
nit: comment missing

embano1 reviewed (commented):
I haven't checked the whole PR because for some parts especially structured/binary handling, I'd like to see more tests for non-happy path scenarios and mis-use if possible. Could you please also add Confluent to the Github integration tests so we run this against Confluent images?

yanmxa commented:
> I haven't checked the whole PR because for some parts especially structured/binary handling, I'd like to see more tests for non-happy path scenarios and mis-use if possible. Could you please also add Confluent to the Github integration tests so we run this against Confluent images?

Thank @embano1! 

I added some tests for the misuse cases when the consumers try to initialize the client. If others come up with scenarios beyond this, we can continually improve it in the future.

Also, I updated the GitHub integration test workflow to add the kafka_confluent service(running on port 9192, the sarama running on 9092) with the confluent image.  So the current confluent kafka protocol uses it for the tests.

embano1 commented on .github/workflows/integration.yaml:
```diff
@@ -27,6 +27,17 @@ jobs:
           - 9092:9092
 
+      kafka_confluent:
+        image: confluentinc/confluent-local:latest
```
note to myself: Ideally we should always use tags (or better) SHAs. But this is not done in the other workflows either yet, making troubleshooting harder.

yanmxa replied:
Done with `latest` -> `7.6.0`.

embano1 commented on protocol/kafka_confluent/v2/message.go:
```diff
@@ -0,0 +1,146 @@
+	// add the kafka message key, topic, partition and partition key to the properties
+	properties[prefix+KafkaOffsetKey] = []byte(strconv.FormatInt(int64(msg.TopicPartition.Offset), 10))
+	properties[prefix+KafkaPartitionKey] = []byte(strconv.FormatInt(int64(msg.TopicPartition.Partition), 10))
+	if msg.TopicPartition.Topic == nil {
+		panic("the kafka message topic shouldn't be nil")
+	}
```
note: the nil check should come at the beginning of the function and check whether:

- `msg` is nil
- `msg.TopicPartition.Topic` is nil

yanmxa replied:
Done.

embano1 commented on protocol/kafka_confluent/v2/message_test.go:
```diff
@@ -0,0 +1,113 @@
+		TopicPartition: topicPartition,
+		Value:          []byte("hello world!"),
+		Headers: mustToKafkaHeaders(map[string]string{
+			"ce_type":            testEvent.Type(),
+			"ce_source":          testEvent.Source(),
+			"ce_id":              testEvent.ID(),
+			"ce_time":            test.Timestamp.String(),
+			"ce-specversion":     "1.0",
+			"ce_dataschema":      test.Schema.String(),
+			"ce_datacontenttype": "text/json",
+			"ce_subject":         "receiverTopic",
```
wondering why they're all using `_` instead of `-`?

yanmxa replied:
Nice catch! 
It's a typo. I change the `_` into `-`, to align them with the `prefix` defined in `message.go`

embano1 commented on protocol/kafka_confluent/v2/message_test.go:
```diff
@@ -0,0 +1,113 @@
+			"ce_dataschema":      test.Schema.String(),
+			"ce_datacontenttype": "text/json",
+			"ce_subject":         "receiverTopic",
+			"ce_exta":            "someext",
```
question: for custom extensions, generally we don't use `ce` prefix. Is this supposed to test custom extension?

embano1 commented on protocol/kafka_confluent/v2/message_test.go:
```diff
@@ -0,0 +1,113 @@
+	}
+}
+
+func mustToKafkaHeaders(m map[string]string) []kafka.Header {
```
nit: `must` functions usually panic

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,178 @@
+var offsetKey = topicPartitionOffsetsType{}
+
+// WithTopicPartitionOffsets will set the positions where the consumer starts consuming from. This option is not required.
+func WithTopicPartitionOffsets(ctx context.Context, topicPartitionOffsets []kafka.TopicPartition) context.Context {
```
nice!

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,178 @@
+	return context.WithValue(ctx, offsetKey, topicPartitionOffsets)
+}
+
+// TopicPartitionOffsetsFrom looks in the given context and returns `[]kafka.TopicPartition`
```
```suggestion
// TopicPartitionOffsetsFrom looks in the given context and returns []kafka.TopicPartition or nil if not set
```

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,178 @@
+var topicPartitionKey = topicPartitionKeyType{}
+
+// WithTopicPartition returns back a new context with the given partition. This option is not required.
+func WithTopicPartition(ctx context.Context, partition int32) context.Context {
```
Wondering where/how this would be used? For example, there's also `WithTopicPartitionsOffsets` which I found similar and the latter one referencing a topic whereas this function just takes a partition (for which topic)?

embano1 commented on samples/kafka_confluent/go.mod:
```diff
@@ -0,0 +1,23 @@
+
+replace github.com/cloudevents/sdk-go/v2 => ../../v2
+
+replace github.com/cloudevents/sdk-go/protocol/kafka_confluent/v2 => ../../protocol/kafka_confluent/v2
```
needed?

yanmxa replied:
yes, we need it

embano1 commented on protocol/kafka_confluent/v2/message.go:
```diff
@@ -0,0 +1,146 @@
+		}
+	}
+
+	if m.internal.Value != nil {
+		err = encoder.SetData(bytes.NewBuffer(m.internal.Value))
+	}
```
btw: have you tested (E2E) with tombstone records, i.e., where value is `null` (`nil`)? Curious about the behavior.

yanmxa replied:
The value with a `nil` case is covered in the integration test. 
The input(sender) event:
```
Context Attributes,
  specversion: 1.0
  type: com.example.MinEvent
  source: http://example.com/source
  id: min-event-2
```

the output(receiver) event:
```bash
Context Attributes,
  specversion: 1.0
  type: com.example.MinEvent
  source: http://example.com/source
  id: min-event-2
Extensions,
  kafkaoffset: 0
  kafkapartition: 0
  kafkatopic: test-ce-confluent-56b0d978-1e0f-4aa3-aecd-993322685530
```

embano1 commented on .github/workflows/integration.yaml:
```diff
@@ -27,6 +27,17 @@ jobs:
 
+      kafka_confluent:
+        image: confluentinc/confluent-local:latest
+        hostname: broker
+        container_name: broker
```
tests are failing due to this; needed?

yanmxa replied:
Yes. I tried to remove it but failed if we set a customized port(9192, the default is 9092) to expose the service on the workflow

embano1 replied:
Please fix, otherwise can't approve

yanmxa replied:
@embano1 I have removed the `hostname` and `container_name` from here now.

yanmxa commented:
@embano1 
Thanks for your patient reviews all along!
Cause several multi-cloud management projects in our team depend on this. If you could provide feedback and suggestions on this PR as soon as possible, it would be of great help to us!

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,212 @@
+	}
+
+	if p.kafkaConfigMap == nil && p.producer == nil && p.consumer == nil {
+		return nil, fmt.Errorf("at least one of the following to initialize the protocol: config, producer, or consumer")
```
```suggestion
		return nil, fmt.Errorf("at least one of the following to initialize the protocol must be set: config, producer, or consumer")
```

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,212 @@
+			p.producer = producer
+		}
+		if p.producer == nil && p.consumer == nil {
+			return nil, fmt.Errorf("at least set one of the receiver or sender topic")
```
```suggestion
			return nil, fmt.Errorf("at least receiver or sender topic must be set")
```

embano1 commented on protocol/kafka_confluent/v2/option.go:
```diff
@@ -0,0 +1,151 @@
+
+var keyForMessageKey = messageKeyType{}
+
+// WithMessageKey returns back a new context with the given messageKey. This option is not required.
```
```suggestion
// WithMessageKey returns back a new context with the given messageKey.
```

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,212 @@
+
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.producer == nil {
+		return fmt.Errorf("the producer client must not be nil")
```
nit (same for others:) 

```suggestion
		return fmt.Errorf("producer client must be set")
```

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return fmt.Errorf("the consumer client must not be nil")
+	}
+	if p.consumerTopics == nil {
+		return fmt.Errorf("the consumer topics must not be nil")
+	}
```
see comment above

embano1 commented on samples/kafka_confluent/receiver-assign/main.go:
```diff
@@ -0,0 +1,60 @@
+	ctx, cancel := context.WithCancel(context.Background())
+
+	receiver, err := kafka_confluent.New(kafka_confluent.WithConfigMap(&kafka.ConfigMap{
+		"bootstrap.servers": "127.0.0.1:9092",
```
will confluent bind to `9092` or `9091` by default?

yanmxa replied:
The confluent binds to 9092 by default

embano1 commented on samples/kafka_confluent/receiver-assign/main.go:
```diff
@@ -0,0 +1,60 @@
+	kafka_confluent "github.com/cloudevents/sdk-go/protocol/kafka_confluent/v2"
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/client"
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
```
nit: if you use import alias `confluent` code will be `confluent.New` and easier to read

yanmxa replied:
Done

embano1 commented on samples/kafka_confluent/receiver-assign/main.go:
```diff
@@ -0,0 +1,60 @@
+
+	defer receiver.Close(ctx)
+
+	c, err := cloudevents.NewClient(receiver, client.WithPollGoroutines(1))
```
question: reason you set goroutines to `1`?

yanmxa replied:
I think that using multiple goroutines may break the order of messages in the Kafka partition, please correct me if I am wrong.

embano1 replied:
Hm, ideally you'd have `WithPollGoroutines == kafka_partitions`, right?

yanmxa replied:
Emm I don't think so. It's a little bit complicated.

Cause Kafka uses multiple goroutines(consumer group) to poll messages from the partitions on the broker. 
Ideally, the `max number of the consumer instance in the group` = the `number of the Kafka topic partitions`, and each of consumer will get a unique message from that assigned partition. This allows Kafka consumption throughput to reach its maximum while ensuring local ordering (partition). All these features all provided by Kafka itself. 

From my current understanding, Cloudevents also uses a multi-goroutines model to improve receiver consumption speed. However, if multiple `PollGoroutines` are used in conjunction with Kafka's consumer group feature, it will disrupt the intended purpose of the Kafka consumer group model because we cannot guarantee message orderliness under the message key partitioning mechanism.

So I set the `WithPollGoroutines` to `1` based on the above consideration

embano1 replied:
Just to check: you're saying the Confluent implementation handles goroutines per consumer group accordingly by default?

yanmxa replied:
Yes, by default, the consumer group associated with the confluent consumer manages all consumers, whether they are implemented as multiple goroutines or multiple threads.

embano1 commented on samples/kafka_confluent/receiver-assign/main.go:
```diff
@@ -0,0 +1,60 @@
+	} else {
+		log.Printf("receiver stopped\n")
+	}
+	cancel()
```
move up to ctx creation and use `defer cancel()`

yanmxa replied:
Done

embano1 replied:
and please remove this here

yanmxa replied:
Done

embano1 reviewed (changes requested):
Some minor things left, sorry...had some time to dig a bit deeper. Should be easy to address. Then please squash your commits.

yanmxa commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,213 @@
+			case kafka.Error:
+				// Errors should generally be considered informational, the client will try to automatically recover.
+				// But in here, we choose to terminate the application if all brokers are down.
+				logger.Infof("Error %v: %v", e.Code(), e)
```
It has the `Errorf`. 
But I have read the [confluent official samples](https://github.com/confluentinc/confluent-kafka-go/blob/master/examples/producer_example/producer_example.go), then all treat this `kafka.Error` as an informational level. 
So I added an optional callback function named `ErrorHandler`. The customer can choose to proceed or abort the message consuming based on the error type.

embano1 replied:
OK

yanmxa commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,213 @@
+			switch e := ev.(type) {
+			case *kafka.Message:
+				p.incoming <- e
+			case kafka.Error:
```
Yes. The `kafka.Error` shouldn't be a pointer. Otherwise, we will not catch the error message on the switch case.

embano1 replied:
Got it, wasn't sure whether they use pointers here as well.

yanmxa commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,213 @@
+	for run {
+		select {
+		case <-ctx.Done():
+			run = false
```
Done

yanmxa commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,213 @@
+					run = false
+				}
+			default:
+				fmt.Printf("Ignored %v\n", e)
```
Done

yanmxa commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	logger.Infof("Closing consumer %v", p.consumerTopics)
+	return p.consumer.Close()
```
Nice catch! 
1. It returns the error message in the for loop now.
2. If someone calls `Close()` concurrently. It will correctly handle this.

yanmxa commented:
> Some minor things left, sorry...had some time to dig a bit deeper. Should be easy to address. Then please squash your commits.

Never mind! 
Done.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+	consumerTopics       []string
+	consumerRebalanceCb  kafka.RebalanceCb     // optional
+	consumerPollTimeout  int                   // optional
+	consumerErrorHandler func(err kafka.Error) //optional
```
regarding error handling discussion: let's add `ctx` here so users can shove in custom loggers without having to use closures?

yanmxa replied:
Done.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+			p.producer = producer
+		}
+		if p.producer == nil && p.consumer == nil {
+			return nil, fmt.Errorf("at least receiver or sender topic must be set")
```
sorry, just saw: any linter would flag this at some point: `Errorf` without variable placeholders (same with other errors)

yanmxa replied:
Done.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+					p.consumerErrorHandler(e)
+				}
+				if e.Code() == kafka.ErrAllBrokersDown {
+					logger.Debugf("All broker connections are down!")
```
so let's make this `Errorf` then to not hide this error when debugging is off.

yanmxa replied:
Done.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+					return e
+				}
+			default:
+				logger.Debugf("Ignored the kafka.Event: %v", e)
```
question: will this print the pointer or real value? What could `e` be as an example?

embano1 replied:
(asking perhaps we don't need `default` at all)

yanmxa replied:
Well, actually I haven't come across this default case. This code is written according to the [official samples](https://github.com/confluentinc/confluent-kafka-go/blob/f1230c0e9dd4b038986ecf26a901e945ec71ca23/examples/consumer_example/consumer_example.go#L117-L118). I am not sure if it contains some cases we unexpected. 

I agree with your idea and have removed the `default`. We can add it back if it is really needed in the future.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+func (p *Protocol) Close(ctx context.Context) error {
+	if p.consumer != nil && !p.consumer.IsClosed() {
```
Now you have a data race, if this is called from another goroutine. You need to protect with mutex here as well and be careful with deadlocking

yanmxa replied:
I added another lock(mutex) to handle the data race between the `Opener` and `Closer`. It will make sure that the `p.consumer.Close()` and `p.consumer().IsClose()` will don't execute in parallel.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+		return fmt.Errorf("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
```
can lead to deadlocks since you're holding on it for function lifetime. See comment below

yanmxa replied:
Whether the mutex is intended to protect the `OpenInbound` can only be invoked once to avoid launching the consumer multiple times. Cause the saram implementation also uses this mutex for the "consumer loop" on the `OpenInbound`.

embano1 replied:
Ah, got it yes.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+	}
+
+	defer func() {
+		if !p.consumer.IsClosed() {
```
use mutex here when removing the "global" mutex above

yanmxa replied:
See the [above](https://github.com/cloudevents/sdk-go/pull/988#discussion_r1527525977).

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+		}
+	}()
+
+	for !p.consumer.IsClosed() {
```
simply drop this bool and make it a simple `for { select { } }`. We rely on `ctx` to terminate this loop. Just make sure to protext access to `isClosed` with mutex.

yanmxa replied:
Sure! Only keeping the `ctx` to terminate the loop is going to make the logic simpler. However, there is a scenario I'm not sure whether we need to consider:

1. Use the `kafka.Consumer` to initialize a `cloudevents.Client`.  
2. Start polling the message by invoking the `StartReceiver`(blocking) of the `cloudevents.Client`. 
3. If we call the `kafka.Consumer.Close()` outside, then the code will panic causing we can't use a closed consumer to poll message.

Since the `cloudevents.Client` hasn't exposed the `Close()` method, maybe we don't need to handle it? I'm not sure, need your confirm...

yanmxa replied:
And I have removed the `p.consumer.IsClosed()` from the consumer for loop now

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+		select {
+		case <-ctx.Done():
+			logger.Debugf("The consumer context is canceled!")
+			return nil
```
nit: `return ctx.Err()` now.

yanmxa replied:
Done

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,225 @@
+			}
+		}
+	}
+	return fmt.Errorf("the consumer is closed")
```
remove

yanmxa replied:
Done

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
```
question: will this always be this type or do we need assertion check?

yanmxa replied:
I think the answer is yes based on the [official example](https://github.com/confluentinc/confluent-kafka-go/blob/32f065e0ede50a539847a4e9bf2d8b9e8f8f5447/examples/producer_custom_channel_example/producer_custom_channel_example.go#L80-L101) they provide.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
```
is TopicPartition also pointer? nil check required as well?

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	p.closerMux.Lock()
+	p.consumerCtx, p.consumerCancel = context.WithCancel(ctx)
```
also call defer p.consumerCancel() ?

yanmxa replied:
Done!

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+		p.consumerCancel()
+	}
+	if p.consumerIncoming != nil {
+		close(p.consumerIncoming)
```
this can cause a panic in openinbound loop when you're within the poll routine but concurrently close this channel
my recommendation is to not close channels here but handle closing those in send/openinbound and just call consumerCancel() here

yanmxa replied:
Done!

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+		close(p.consumerIncoming)
+	}
+
+	if p.producer != nil && !p.producer.IsClosed() {
```
IMHO race with Send

yanmxa replied:
Yes! I also leverage the `closerMux` to handle the race condition between `Sender` and `Closer`

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+		p.producer.Close()
+	}
+	if p.producerDeliveryChan != nil {
+		close(p.producerDeliveryChan)
```
can cause panic if concurrently in send, see comment above

yanmxa replied:
Done.

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.producer == nil {
```
if you add a producerMtx check here whether it's "closed" you can return early and avoid concurrency issues (see comments in Close())

yanmxa replied:
The `closerMux` is intended to handle the race condition between `Sender`/`Opener` and `Closer`. So instead add the new mutex(producerMtx), I still use the `closerMux` to handle it.

embano1 reviewed (changes requested):
Looked closer at the Close/Send/OpenInbound implementations and there's still races/chances for panics. Left comments on how to simplify

yanmxa commented:
> Looked closer at the Close/Send/OpenInbound implementations and there's still races/chances for panics. Left comments on how to simplify

Thanks a lot!
Done!

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,249 @@
+	for {
+		select {
+		case <-p.consumerCtx.Done():
+			return fmt.Errorf("the consumer context is canceled: %v", p.consumerCtx.Err())
```
return ctx.Err()

yanmxa replied:
@embano1 Done.

embano1 commented:
@yanmxa 

I spend some time looking at the `OpenInbound/Send/Receive/Close` functions and suggesting this patch to remove some unneeded code and tidy things up:

```patch
diff --git a/protocol/kafka_confluent/v2/protocol.go b/protocol/kafka_confluent/v2/protocol.go
index 249855c..7c83087 100644
--- a/protocol/kafka_confluent/v2/protocol.go
+++ b/protocol/kafka_confluent/v2/protocol.go
@@ -99,10 +99,8 @@ func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ..
 
 	p.closerMux.Lock()
 	defer p.closerMux.Unlock()
+	
 	if p.producer.IsClosed() {
-		if p.producerDeliveryChan != nil && !isClosed(p.producerDeliveryChan) {
-			close(p.producerDeliveryChan)
-		}
 		return errors.New("producer is closed")
 	}
 
@@ -177,9 +175,7 @@ func (p *Protocol) OpenInbound(ctx context.Context) error {
 				logger.Errorf("failed to close the consumer: %v", err)
 			}
 		}
-		if p.consumerIncoming != nil && !isClosed(p.producerDeliveryChan) {
-			close(p.consumerIncoming)
-		}
+		close(p.consumerIncoming)
 	}()
 
 	for {
@@ -228,21 +224,14 @@ func (p *Protocol) Close(ctx context.Context) error {
 	p.closerMux.Lock()
 	defer p.closerMux.Unlock()
 
-	if p.consumerCancel != nil {
-		p.consumerCancel()
-	}
-
 	if p.producer != nil && !p.producer.IsClosed() {
 		p.producer.Close()
+		close(p.producerDeliveryChan)
 	}
-	return nil
-}
 
-func isClosed(ch <-chan kafka.Event) bool {
-	select {
-	case <-ch:
-		return true
-	default:
+	if p.consumerCancel != nil {
+		p.consumerCancel()
 	}
-	return false
+
+	return nil
 }
```

embano1 commented:
Let me know if you have questions.

yanmxa commented:
> Let me know if you have questions.

Thanks! 
Your advice makes the code clearer and no other questions.

embano1 commented:
> I have a slight concern, perhaps unnecessary and we can just ignore it. That is what if the `producer` isn't closed by the `Closer`, that will cause `producerDeliveryChan` not to be closed?

Please add a comment to the `Close()` method that protocol users must call `Close()` to properly clean up resources after use.

yanmxa commented:
> > I have a slight concern, perhaps unnecessary and we can just ignore it. That is what if the `producer` isn't closed by the `Closer`, that will cause `producerDeliveryChan` not to be closed?
> 
> Please add a comment to the `Close(ctx)` method that protocol users must call `Close(ctx)` to properly clean up resources after use. Also, add `defer p.Close(ctx)` to all your examples

Done!

embano1 commented:
Please add a doc string to the `Close()` method as well, advising users to call `Close()` after use. Please directly squash your commits so we can merge. Also, any related documentation/README updates needed to further highlight this new protocol implementation?

embano1 commented:
Also, why did you not use `WithPollGoroutines` in all `receiver` examples? Should be clear why this is advised e.g., also adding a comment to the example explaining that this won't have perf impact and is only required if the user cares about in-order processing when using the SDK.

yanmxa commented:
> Also, why did you not use `WithPollGoroutines` in all `receiver` examples? Should be clear why this is advised e.g., also adding a comment to the example explaining that this won't have perf impact and is only required if the user cares about in-order processing when using the SDK.

Thanks for your prompt feedback!

1. Add some comments for the `Close` method and `WithPollGoroutines` in all `receiver` examples. 
2. Update the related document in this repository.
3. Squash the commits

Could you please help me to check if the wording in the above document is appropriate?

embano1 commented on protocol/kafka_confluent/v2/protocol.go:
```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+// Please use this method to clean up resources after utilizing the protocol
```
Go comments must follow specific pattern

```suggestion
// Close cleans up resources after use. Must be called to properly close underlying Kafka resources and avoid resource leaks
```

yanmxa replied:
Thanks! https://tip.golang.org/doc/comment

embano1 reviewed (changes requested):
Just that one comment please

embano1 commented:
Can you please check your integration tests? They're not looking positive, but passing: https://github.com/cloudevents/sdk-go/actions/runs/8420863202/job/23056453953?pr=988#step:5:284 (example)

embano1 commented:
This unit test is also not looking ok: https://github.com/cloudevents/sdk-go/actions/runs/8420863203/job/23056453818#step:4:3883

yanmxa commented:
> This unit test is also not looking ok: https://github.com/cloudevents/sdk-go/actions/runs/8420863203/job/23056453818#step:4:3883

The error log disappears in the unit test now.

yanmxa commented:
> Can you please check your integration tests? They're not looking positive, but passing: https://github.com/cloudevents/sdk-go/actions/runs/8420863202/job/23056453953?pr=988#step:5:284 (example)

Done!
The integration test error log is from [here](https://github.com/cloudevents/sdk-go/pull/988/files#diff-82e585cd3911942b03502d3dc38e88ede560ecc00b6615ce9d89b25309099095R203).

embano1 reviewed (approved):
LGTM

THX @yanmxa for getting this over the line!

cc/ @duglin for final overall check

embano1 commented:
@duglin can you please take a final look just overall whether we missed a README or general requirements for a new protocol binding?

duglin commented:
@lionelvillard too :-)

duglin commented on .github/workflows/integration.yaml:
```diff
@@ -27,6 +27,15 @@ jobs:
           - 9091:9091
           - 9092:9092
 
+      kafka_confluent:
```
Should we rename the one above from "kafka" to "kafka_sarama" to be consistent with the rest of the naming?

duglin commented:
LGTM

I'll let @embano1 hit the button.
